packages/amplify-environment-parameters/src/environment-metadata-manager.ts

```diff
@@ -0,0 +1,296 @@
+        });
+      }
+    });
+
+    const optionalKeys: (keyof IEnvironmentMetadata)[] = [
+      'PermissionsBoundaryPolicyArn',
+      'AmplifyAppId',
+    ];
+
+    optionalKeys.forEach(optionalKey => {
+      const typeOfValue = typeof amplifyMeta[optionalKey];
+      if (typeOfValue !== 'string' && typeOfValue !== 'undefined') {
+        throw amplifyFaultWithIssueReportLink('ConfigurationFault', {
+          message: `Tried to initialize EnvironmentMetadata object with ${optionalKey} set to a non-string value`,
+        });
+      }
+    });
+    const validatedAmplifyMeta = amplifyMeta as IEnvironmentMetadata;
+    this.AuthRoleName = validatedAmplifyMeta.AuthRoleName;
+    this.AuthRoleArn = validatedAmplifyMeta.AuthRoleArn;
+    this.UnauthRoleArn = validatedAmplifyMeta.UnauthRoleArn;
+    this.UnauthRoleName = validatedAmplifyMeta.UnauthRoleName;
+    this.Region = validatedAmplifyMeta.Region;
+    this.DeploymentBucketName = validatedAmplifyMeta.DeploymentBucketName;
+    this.StackId = validatedAmplifyMeta.StackId;
+    this.StackName = validatedAmplifyMeta.StackName;
+    this._AmplifyAppId = validatedAmplifyMeta.AmplifyAppId;
+    this._PermissionsBoundaryPolicyArn = validatedAmplifyMeta.PermissionsBoundaryPolicyArn;
+    this._dirty = isDirty;
+  }
+
+  get PermissionsBoundaryPolicyArn(): string | undefined {
+    return this._PermissionsBoundaryPolicyArn || getPermissionsBoundaryArn();
+  }
+
+  set PermissionsBoundaryPolicyArn(value: string | undefined) {
+    this._PermissionsBoundaryPolicyArn = value;
+    this._dirty = true;
+  }
+
+  get AmplifyAppId(): string {
+    return this._AmplifyAppId!;
+  }
+
+  set AmplifyAppId(value: string) {
+    this._AmplifyAppId = value;
+    this._dirty = true;
+  }
+
+  /**
+   * Persist the metadata stored in the manager object
+   *
+   * currently this writes the values out to the `amplify-meta.json` file and `team-provider-info.json` but eventually we will git rid of
+   * `team-provider-info.json`. And beyond that the values stored in `amplify-meta.json` should eventually be stored in a metadata service
+   */
+  write(respectIsDirty = true, writeTarget: WriteTarget = localFilesWriteTarget): void {
+    if (respectIsDirty && !this._dirty) {
+      return;
+    }
+    writeTarget(this.toObject());
+  }
+
+  private toObject(): Record<string, string> {
+    const obj = {
+      AuthRoleName: this.AuthRoleName,
+      AuthRoleArn: this.AuthRoleArn,
+      UnauthRoleArn: this.UnauthRoleArn,
+      UnauthRoleName: this.UnauthRoleName,
+      Region: this.Region,
+      DeploymentBucketName: this.DeploymentBucketName,
+      StackName: this.StackName,
+      StackId: this.StackId,
+      AmplifyAppId: this._AmplifyAppId,
+    } as Record<string, string>;
+    if (this.PermissionsBoundaryPolicyArn) {
+      obj.PermissionsBoundaryPolicyArn = this.PermissionsBoundaryPolicyArn;
+    }
```

Comment:
why do we have to convert this to anonymous object and carve out this property conditionally?
is there a way to configure serialization in such a way that it skips null/undef properties ?